<a href="https://colab.research.google.com/github/marcosvnespolo/Portifolio/blob/main/Avalia%C3%A7%C3%A3o_de_Vinhos_com_ML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Overview do Projeto


*   O objetivo do projeto é a utilização de algorítimos de ML para classificação de vinhos, identificando quais características tornam um vinho tinto ou branco;
*   A base de dadas utilizada pode ser encontrada em: https://www.kaggle.com/datasets/dell4010/wine-dataset.




In [2]:
import pandas as pd
dataset = pd.read_csv('/content/drive/MyDrive/Datasets/Avaliação de Vinhos/wine_dataset.csv')
display(dataset)

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0               7.4              0.70         0.00             1.9      0.076   
1               7.8              0.88         0.00             2.6      0.098   
2               7.8              0.76         0.04             2.3      0.092   
3              11.2              0.28         0.56             1.9      0.075   
4               7.4              0.70         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
6492            6.2              0.21         0.29             1.6      0.039   
6493            6.6              0.32         0.36             8.0      0.047   
6494            6.5              0.24         0.19             1.2      0.041   
6495            5.5              0.29         0.30             1.1      0.022   
6496            6.0              0.21         0.38             0.8      0.020   

      free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
6492                 24.0                  92.0  0.99114  3.27       0.50   
6493                 57.0                 168.0  0.99490  3.15       0.46   
6494                 30.0                 111.0  0.99254  2.99       0.46   
6495                 20.0                 110.0  0.98869  3.34       0.38   
6496                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  style  
0         9.4        5    red  
1         9.8        5    red  
2         9.8        5    red  
3         9.8        6    red  
4         9.4        5    red  
...       ...      ...    ...  
6492     11.2        6  white  
6493      9.6        5  white  
6494      9.4        6  white  
6495     12.8        7  white  
6496     11.8        6  white  

[6497 rows x 13 columns]

Para proceder com as análises, é necessário que a coluna 'Style' seja composta apenas por valores numéricos

In [3]:
dataset['style'] = dataset['style'].replace('red', 0)
dataset['style'] = dataset['style'].replace('white', 1)

In [4]:
display(dataset)

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0               7.4              0.70         0.00             1.9      0.076   
1               7.8              0.88         0.00             2.6      0.098   
2               7.8              0.76         0.04             2.3      0.092   
3              11.2              0.28         0.56             1.9      0.075   
4               7.4              0.70         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
6492            6.2              0.21         0.29             1.6      0.039   
6493            6.6              0.32         0.36             8.0      0.047   
6494            6.5              0.24         0.19             1.2      0.041   
6495            5.5              0.29         0.30             1.1      0.022   
6496            6.0              0.21         0.38             0.8      0.020   

      free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
6492                 24.0                  92.0  0.99114  3.27       0.50   
6493                 57.0                 168.0  0.99490  3.15       0.46   
6494                 30.0                 111.0  0.99254  2.99       0.46   
6495                 20.0                 110.0  0.98869  3.34       0.38   
6496                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  style  
0         9.4        5      0  
1         9.8        5      0  
2         9.8        5      0  
3         9.8        6      0  
4         9.4        5      0  
...       ...      ...    ...  
6492     11.2        6      1  
6493      9.6        5      1  
6494      9.4        6      1  
6495     12.8        7      1  
6496     11.8        6      1  

[6497 rows x 13 columns]

Podemos separar as variáveis em preditora e variável Avaliação

In [5]:
y = dataset['style']
x = dataset.drop('style', axis =1)
display(x, y)

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0               7.4              0.70         0.00             1.9      0.076   
1               7.8              0.88         0.00             2.6      0.098   
2               7.8              0.76         0.04             2.3      0.092   
3              11.2              0.28         0.56             1.9      0.075   
4               7.4              0.70         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
6492            6.2              0.21         0.29             1.6      0.039   
6493            6.6              0.32         0.36             8.0      0.047   
6494            6.5              0.24         0.19             1.2      0.041   
6495            5.5              0.29         0.30             1.1      0.022   
6496            6.0              0.21         0.38             0.8      0.020   

      free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
6492                 24.0                  92.0  0.99114  3.27       0.50   
6493                 57.0                 168.0  0.99490  3.15       0.46   
6494                 30.0                 111.0  0.99254  2.99       0.46   
6495                 20.0                 110.0  0.98869  3.34       0.38   
6496                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
3         9.8        6  
4         9.4        5  
...       ...      ...  
6492     11.2        6  
6493      9.6        5  
6494      9.4        6  
6495     12.8        7  
6496     11.8        6  

[6497 rows x 12 columns]

0       0
1       0
2       0
3       0
4       0
       ..
6492    1
6493    1
6494    1
6495    1
6496    1
Name: style, Length: 6497, dtype: int64

Separando os conjustos de dados de treino e de teste

In [6]:
from sklearn.model_selection import train_test_split
x_treino, x_teste, y_treino, y_teste = train_test_split(x, y, test_size = 0.3)

Criando o modelo de ML e imprimindo o resultado

In [7]:
from sklearn.ensemble import ExtraTreesClassifier
modelo = ExtraTreesClassifier()
modelo.fit(x_treino, y_treino)

resultado = modelo.score(x_teste, y_teste)
print('Acurácia: ', resultado)

Acurácia:  0.9958974358974358




*   A acurácia compara os resultados reais com os obtidos pelo mdelo;
*   Nesete caso, o modelo se demonstrou bastante satisfatório, acertando a previsão em praticamente todas as tentativas.



In [8]:
y_teste[400:420]

2219    1
3958    1
1869    1
1538    0
1055    0
992     0
2154    1
3270    1
4503    1
715     0
6141    1
4165    1
2190    1
1671    1
982     0
1075    0
2799    1
466     0
4167    1
5503    1
Name: style, dtype: int64

In [9]:
x_teste[400:420]

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
2219            6.5             0.460         0.41            16.8      0.084   
3958            6.0             0.495         0.27             5.0      0.157   
1869            6.9             0.320         0.26             8.3      0.053   
1538            6.2             0.640         0.09             2.5      0.081   
1055            8.2             0.640         0.27             2.0      0.095   
992             6.5             0.400         0.10             2.0      0.076   
2154            9.2             0.250         0.34             1.2      0.026   
3270            7.0             0.350         0.30             6.5      0.028   
4503            6.4             0.250         0.32             0.9      0.034   
715             7.2             0.490         0.18             2.7      0.069   
6141            6.9             0.280         0.37             9.1      0.037   
4165            6.9             0.210         0.24             1.8      0.021   
2190            7.3             0.200         0.37             1.2      0.037   
1671            6.0             0.270         0.28             4.8      0.063   
982             7.3             0.520         0.32             2.1      0.070   
1075            9.1             0.250         0.34             2.0      0.071   
2799            6.8             0.180         0.37             1.5      0.027   
466            10.3             0.500         0.42             2.0      0.069   
4167            6.8             0.280         0.36             7.0      0.043   
5503            5.0             0.455         0.18             1.9      0.036   

      free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
2219                 59.0                 222.0  0.99930  3.18       0.58   
3958                 17.0                 129.0  0.99396  3.03       0.36   
1869                 32.0                 180.0  0.99650  3.25       0.51   
1538                 15.0                  26.0  0.99538  3.57       0.63   
1055                  5.0                  77.0  0.99747  3.13       0.62   
992                  30.0                  47.0  0.99554  3.36       0.48   
2154                 31.0                  93.0  0.99160  2.93       0.37   
3270                 27.0                  87.0  0.99360  3.40       0.42   
4503                 40.0                 114.0  0.99114  3.31       0.58   
715                  13.0                  34.0  0.99670  3.29       0.48   
6141                 16.0                  76.0  0.99480  3.05       0.54   
4165                 17.0                  80.0  0.98992  3.15       0.46   
2190                 48.0                 119.0  0.99200  3.32       0.49   
1671                 31.0                 201.0  0.99640  3.69       0.71   
982                  51.0                  70.0  0.99418  3.34       0.82   
1075                 45.0                  67.0  0.99769  3.44       0.86   
2799                 37.0                  93.0  0.99200  3.30       0.45   
466                  21.0                  51.0  0.99820  3.16       0.72   
4167                 60.0                 207.0  0.99556  3.16       0.49   
5503                 33.0                 106.0  0.98746  3.21       0.83   

      alcohol  quality  
2219      9.0        5  
3958      9.3        5  
1869      9.2        6  
1538     12.0        5  
1055      9.1        6  
992       9.4        6  
2154     11.3        7  
3270     11.4        7  
4503     10.8        7  
715       9.2        6  
6141     11.1        5  
4165     12.3        7  
2190     10.9        6  
1671     10.0        5  
982      12.9        6  
1075     10.2        7  
2799     10.8        6  
466      11.5        6  
4167      9.6        6  
5503     14.0        7

In [11]:
previsao = modelo.predict(x_teste[400:420])
previsao

array([1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1])

## Conclusões


*   A precisão do modelo foi de aproximadamente 99.6%, o que significa que o mesmo nos retornará um resultado correto em praticamente todas as tentativas;
*  Nas últimas linhas de código foi realizada uma prova real da acurácia do modelo, o que nos confirma que o mesmo está nos retornando resultados satisfatórios e condizentes com os reais na maioria das tentativas.


